File: leet48/leet48.cpp
// https://leetcode.com/problems/rotate-image/

#include <vector>
using namespace std;

class Solution
{
public:
    void transpose(vector<vector<int>> &matrix, int n)
    {
        for (int i = 0; i < n; i++)
        {
            for (int j = i + 1; j < n; j++)
            {
                matrix[i][j] = matrix[i][j] ^ matrix[j][i];
                matrix[j][i] = matrix[i][j] ^ matrix[j][i];
                matrix[i][j] = matrix[i][j] ^ matrix[j][i];
            }
        }
    }

    void rotate(vector<vector<int>> &matrix)
    {
        int n = matrix[0].size();
        transpose(matrix, n);
        for (int i = 0; i < n; i++)
        {
            reverse(matrix[i].begin(), matrix[i].end());
        }
    }
};

File: leet48/leet48.py
# https://leetcode.com/problems/rotate-image/

from typing import List

class Solution:
    def transpose(self, matrix: List[List[int]], n: int) -> None:
        for i in range(n):
            for j in range(i + 1, n):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]

    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        n = len(matrix)
        self.transpose(matrix, n)
        for i in range(n):
            matrix[i][:] = matrix[i][::-1]
